manim/utils/decorators.py

```diff
@@ -0,0 +1,445 @@
+"""Function decorators."""
+
+__all__ = ["deprecated", "deprecated_params"]
+
+
+import re
+from typing import Any, Callable, Iterable, List, Optional, Tuple, Union
+
+from decorator import decorate, decorator
+
+from .. import logger
+
+
+def get_callable_info(callable: Callable) -> Tuple[str, str]:
+    """Returns type and name of a callable.
+
+    Parameters
+    ----------
+    callable
+        The callable
+
+    Returns
+    -------
+    Tuple[str, str]
+        The type and name of the callable. Type can can be one of "class", "method" (for
+        functions defined in classes) or "function"). For methods, name is Class.method.
+    """
+    what = type(callable).__name__
+    name = callable.__qualname__
+    if what == "function" and name[0].isupper():  # TODO a bit hacky but works
+        what = "method"
+    elif what == "type":
+        what = "class"
+    return (what, name)
```

Comment:
Rather than accessing the internal methods `__qualname__` I think it should be better to use https://docs.python.org/3/library/types.html#module-types

For finding whether is a method you can check whether both `types.MethodType` and the `callable` are of the same instance using `isinstance`. For functions, it would `types.FunctionType`.

Reply:
> For finding whether is a method you can check whether both `types.MethodType` and the callable are of the same instance using `isinstance`.

What you mean is `isinstance(callable, MethodType)`,. right? This doesn't work. The reason is that python only considers functions of already instantiated objetcs as methods (`Foo().bar` is a method, `Foo.bar` not).

Reply:
I just found this on stackoverflow: https://stackoverflow.com/questions/63455475/how-to-test-if-a-decorator-is-decorating-a-method-or-a-function

Considering that it is impossible I quite like my approach :smile: